components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/HSMBasedExternalCryptoProvider.java

```diff
@@ -0,0 +1,534 @@
+        Session session = initiateSession(slotResolver.resolveSlot(cryptoContext), false);
+        PrivateKey retrievedKey;
+        try {
+            retrievedKey = (PrivateKey) retrieveKey(privateKeyTemplate, session);
+        } finally {
+            sessionHandler.closeSession(session);
+        }
+        if (!retrievedKey.getSensitive().getBooleanValue() && retrievedKey.getExtractable().getBooleanValue()) {
+            return PKCS11JCEObjectMapper.mapPrivateKeyPKCS11ToJCE(retrievedKey);
+        } else {
+            String errorMessage = String.format("Requested private key %s is not extractable.",
+                    privateKeyInfo.getKeyAlias());
+            throw new CryptoException(errorMessage);
+        }
+    }
+
+    /**
+     * Computes and returns the {@link HybridEncryptionOutput} based on provided {@link HybridEncryptionInput}
+     * Hybrid encryption is carried out using the underlying HSM device.
+     * Session object is created for symmetric key for decryption.
+     *
+     * @param hybridEncryptionInput Input data for hybrid encryption.*
+     * @param symmetricAlgorithm    The symmetric encryption/decryption algorithm.
+     * @param asymmetricAlgorithm   The asymmetric encryption/decryption algorithm.
+     * @param javaSecurityProvider  The Java Security API provider. This value is discarded in this component.
+     * @param cryptoContext         The context information which is used to discover
+     *                              the public key of the external entity.
+     * @return {@link HybridEncryptionOutput} cipher text with required parameters
+     * @throws CryptoException
+     */
+    @Override
+    public HybridEncryptionOutput hybridEncrypt(HybridEncryptionInput hybridEncryptionInput, String symmetricAlgorithm,
+                                                String asymmetricAlgorithm, String javaSecurityProvider,
+                                                CryptoContext cryptoContext, CertificateInfo certificateInfo)
+            throws CryptoException {
+
+        MechanismDataHolder mechanismDataHolder = new MechanismDataHolder(ENCRYPT_MODE, symmetricAlgorithm,
+                hybridEncryptionInput.getAuthData());
+        Mechanism symmetricMechanism = mechanismResolver.resolveMechanism(mechanismDataHolder);
+        Session session = initiateSession(slotResolver.resolveSlot(cryptoContext), true);
+        byte[] encryptedData;
+        SecretKey encryptionKey = getSymmetricKey(session, symmetricAlgorithm, null, true);
+        try {
+            Cipher cipher = new Cipher(session);
+            encryptedData = cipher.encrypt(hybridEncryptionInput.getPlainData(), encryptionKey, symmetricMechanism);
+        } finally {
+            sessionHandler.closeSession(session);
+        }
+        byte[] encryptedKey;
+        if (encryptionKey instanceof AESSecretKey) {
+            encryptedKey = encrypt(((AESSecretKey) encryptionKey).getValue().getByteArrayValue(),
```

Comment:
Do we really need to cast this and send ?

Reply:
Yes ayye, because 'getValue()' method is not defined for SecretKey objects.